[DCOS-29502] Add default user auth mappings to resolve principal

mockingbird2 opened the pull request:
This populates the `hadoop.security.auth_to_local` in the `core-site.xml` with auth user  auth mappings by default. Thus the Hdfs tasks can resolve its principal without the need of additional variables in the environment.

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java:
```diff
@@ -55,28 +57,57 @@ private static SchedulerBuilder createSchedulerBuilder(File yamlSpecFile) throws
+        String userAuthMapping;
+
+        if (Boolean.valueOf(env.get("TASKCFG_ALL_SECURITY_KERBEROS_ENABLED"))) {
+            String frameworkHostname = EndpointUtils.toAutoIpDomain(rawServiceSpec.getName(), schedulerConfig);
```
It might be useful to link this to `FRAMEWORK_HOST` in the variable name?

mockingbird2 replied:
I am not sure what you by this. Isn't the variable name `frameworkHostname` self-explaining?

elezar replied:
I mean that we use `frameworkHost` instead of `frameworkHostname`.

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java:
```diff
@@ -55,28 +57,57 @@ private static SchedulerBuilder createSchedulerBuilder(File yamlSpecFile) throws
+        if (Boolean.valueOf(env.get("TASKCFG_ALL_SECURITY_KERBEROS_ENABLED"))) {
+            String frameworkHostname = EndpointUtils.toAutoIpDomain(rawServiceSpec.getName(), schedulerConfig);
+            HdfsUserAuthMapper authMapper = setupAuthMapper(env, frameworkHostname);
+            authMapper.addUserAuthMappingFromEnv(env, TASKCFG_ALL_AUTH_TO_LOCAL);
```
We could move the `TASKCFG_ALL_AUTH_TO_LOCAL` envvar specification to the `HdfsUserAuthMapper` class instead. Assuming it is not used elsewhere?

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java:
```diff
@@ -55,28 +57,57 @@ private static SchedulerBuilder createSchedulerBuilder(File yamlSpecFile) throws
+            String frameworkHostname = EndpointUtils.toAutoIpDomain(rawServiceSpec.getName(), schedulerConfig);
+            HdfsUserAuthMapper authMapper = setupAuthMapper(env, frameworkHostname);
+            authMapper.addUserAuthMappingFromEnv(env, TASKCFG_ALL_AUTH_TO_LOCAL);
+            authMapper.addDefaultUserAuthMapping(JOURNAL_POD_TYPE, "node", JOURNAL_NODE_COUNT);
```
From my point of view this interface could be simplified. Would it not make sense to have a single function call (or builder) to build the mapping?

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java:
```diff
@@ -55,28 +57,57 @@ private static SchedulerBuilder createSchedulerBuilder(File yamlSpecFile) throws
                 .setPodEnv("name", SERVICE_ZK_ROOT_TASKENV, CuratorUtils.getServiceRootPath(rawServiceSpec.getName()))
-                .setAllPodsEnv(DECODED_AUTH_TO_LOCAL,
-                                getHDFSUserAuthMappings(System.getenv(), TASKCFG_ALL_AUTH_TO_LOCAL))
+                .setAllPodsEnv(DECODED_AUTH_TO_LOCAL, userAuthMapping)
```
I really like how this simplifies!

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java:
```diff
@@ -55,28 +57,57 @@ private static SchedulerBuilder createSchedulerBuilder(File yamlSpecFile) throws
     }
 
+    private static HdfsUserAuthMapper setupAuthMapper(Map<String, String> env, String frameworkHostname) {
+        String realm = env.get("TASKCFG_ALL_SECURITY_KERBEROS_REALM");
```
Would it make sense to move this into the construction of the auth mapper instead?

elezar commented on frameworks/hdfs/src/test/java/com/mesosphere/sdk/hdfs/scheduler/HdfsUserAuthMapperTest.java:
```diff
@@ -0,0 +1,111 @@
+        Assert.assertEquals(findAuthMappings(filledTemplate), mapper.getUserAuthMappingString());
+    }
+
+    private static String findAuthMappings(String config) throws IOException {
```
Although I don't have a problem with handling this here -- it's definitely faster, but we could check this here instead: https://github.com/mesosphere/dcos-commons/blob/master/frameworks/hdfs/tests/test_sanity.py#L61. Would it be possible to use an XML-parser in the test instead? Not critical for this PR though. An alternative is would be to not use the full template in this test.

mockingbird2 replied:
Yes, indeed it would be nice to use a proper XML parsing tool. Unfortunately, I haven't found a good one now. Maybe someone with more Java experience might know a proper solution.

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsUserAuthMapper.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * This class helps to generate user auth mappings for the HDFS setup with kerberos.
+ */
+public class HdfsUserAuthMapper {
```
Should this be `HDFSUserAuthMapper`?

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsUserAuthMapper.java:
```diff
@@ -0,0 +1,47 @@
+        this.frameworkUser = frameworkUser;
+    }
+
+    public void addUserAuthMappingFromEnv(Map<String, String> env, String envVarKeyName) {
```
Add a test for when the specified envvar doesn't exist.

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsUserAuthMapper.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public String getUserAuthMappingString() {
+        return String.join("\n", authMappings);
```
Should we strip blank strings before joining?

mockingbird2 commented:
@elezar thanks for the nice discussion about improving this. I think I addressed most of the feedback regarding the java side. One last thing I have to do is adjusting the `frameworks/hdfs/tests/auth.py` because the default mappings are not needed anymore.

elezar commented on frameworks/hdfs/src/test/java/com/mesosphere/sdk/hdfs/scheduler/HDFSUserAuthMapperBuilderTest.java:
```diff
@@ -60,6 +60,17 @@ public void testAddDefaultUserAuthMapping() {
+                .addUserAuthMappingFromEnv()
+                .addDefaultUserAuthMapping("data", "node", 1)
+                .build();
+        Assert.assertEquals(expected, authMappings.split("\n")[0]);
```
Minor point: This implies a particular ordering. We could do a search in the lines instead.

elezar commented on frameworks/hdfs/tests/auth.py:
```diff
@@ -52,7 +52,7 @@ def get_principal_to_user_mapping() -> str:
     :return: A base64-encoded string of principal->user mappings
     """
-    rules = ["RULE:[2:$1@$0](^hdfs@.*$)s/.*/hdfs/", "RULE:[1:$1@$0](^nobody@.*$)s/.*/nobody/"]
+    rules = ["RULE:[2:$1@$0](^hdfs@.*$)s/.*/hdfs/"]
```
Note that this rule effectively overrides the rules that we define since *ALL* our principals match this pattern.

mockingbird2 replied:
When looking at the actual `core-site.xml` the added rules look like this `hdfs/journal-0-node.testintegrationhdfs.autoip.dcos.thisdcos.directory@LOCAL)s/.*/nobody/`. I already tried removing the other rule as well but then the tests start failing. So I think this rule is an addition.

elezar replied:
If the rules we add with `nobody` don't work, then we need to ensure that our rules look like:
```
RULE:[2:$1/$2@$0](hdfs/journal-0-node.testintegrationhdfs.autoip.dcos.thisdcos.directory@LOCAL)s/.*/hdfs/
```
Remember that this has to map to the HDFS super user which is `hdfs`...

mockingbird2 replied:
Is there anything I have to change? Since the integration test passes without the need for the superuser rule so it is not required for the initial setup at least.

elezar replied:
It may be good to check that this passes on a strict cluster. I'll start a build tomorrow.

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HDFSAuthEnvContainer.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private String getAuthMappingFromEnv(Map<String, String> env) {
+        if (!env.containsKey(TASKCFG_ALL_AUTH_TO_LOCAL)) {
```
Minor: We could also handle the blank string here.

mockingbird2 replied:
I would rather keep it in the builder because the default case for this method is currently to return an empty string. This needs some adjustment if we move the black string logic here.

elezar replied:
OK. That's fine.

For the record, I was thinking of something like:
```java
        if (StringUtils.isBlank(env.getOrDefault("TASKCFG_ALL_AUTH_TO_LOCAL", ""))) {
            return "";
        }
```

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HDFSUserAuthMapperBuilder.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public String build() {
+        authMappings.removeIf(String::isEmpty);
```
Let's use `isBlank` instead.

elezar commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HDFSUserAuthMapperBuilder.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public String build() {
+        authMappings.removeIf(StringUtils::isBlank);
```
minor: Indentation.

Also, are we modifying the state here. What about:
```java
        return authMappings.stream().filter(StringUtils::isBlank).collect(Collectors.joining("\n"));
```
as a single line?

elezar commented:
@mockingbird2 the strict nightly build has passed at least once. Let's merge this and get it backported!
